### ISSUE TITLE
EXC_BAD_ACCESS when running on IOS

### ISSUE THREAD

**Issue by rklerck** [oscarfonts/mapbox-gl-cordova-mbtiles#2]
Hi,

I'm trying to run this on IOS but it fails. I followed the exact same steps for Android but than for IOS but it doesn't work.

In xcode I see the following:

```
2017-09-20 17:49:06.607284+0200 Mapbox GL Cordova MBTiles[9934:3252250] DiskCookieStorage changing policy from 2 to 0, cookie file: file:///private/var/mobile/Containers/Data/Application/AA1920DC-E1D5-4127-9314-15F90EF3CFF6/Library/Cookies/Cookies.binarycookies
2017-09-20 17:49:06.689006+0200 Mapbox GL Cordova MBTiles[9934:3252250] Apache Cordova native platform version 4.4.0 is starting.
2017-09-20 17:49:06.690229+0200 Mapbox GL Cordova MBTiles[9934:3252250] Multi-tasking -> Device: YES, App: YES
2017-09-20 17:49:06.957222+0200 Mapbox GL Cordova MBTiles[9934:3252250] Using UIWebView
2017-09-20 17:49:06.961533+0200 Mapbox GL Cordova MBTiles[9934:3252250] [CDVTimer][handleopenurl] 0.165045ms
2017-09-20 17:49:06.965368+0200 Mapbox GL Cordova MBTiles[9934:3252250] [CDVTimer][intentandnavigationfilter] 3.690958ms
2017-09-20 17:49:06.965680+0200 Mapbox GL Cordova MBTiles[9934:3252250] [CDVTimer][gesturehandler] 0.155985ms
2017-09-20 17:49:06.997877+0200 Mapbox GL Cordova MBTiles[9934:3252250] [CDVTimer][file] 31.947970ms
2017-09-20 17:49:06.999100+0200 Mapbox GL Cordova MBTiles[9934:3252250] [CDVTimer][localstorage] 1.074970ms
2017-09-20 17:49:06.999226+0200 Mapbox GL Cordova MBTiles[9934:3252250] [CDVTimer][TotalPluginStartup] 37.927985ms
2017-09-20 17:49:07.987452+0200 Mapbox GL Cordova MBTiles[9934:3252303] libMobileGestalt MobileGestaltSupport.m:153: pid 9934 (Mapbox GL Cordova MBTiles) does not have sandbox access for frZQaeyWLUvLjeuEK43hmg and IS NOT appropriately entitled
2017-09-20 17:49:07.987599+0200 Mapbox GL Cordova MBTiles[9934:3252303] libMobileGestalt MobileGestalt.c:550: no access to InverseDeviceID (see <rdar://problem/11744455>)
2017-09-20 17:49:09.582580+0200 Mapbox GL Cordova MBTiles[9934:3252250] Resetting plugins due to page load.
2017-09-20 17:49:09.745905+0200 Mapbox GL Cordova MBTiles[9934:3252250] Finished load of: file:///var/containers/Bundle/Application/3C8F2846-8171-4389-A861-58485458CCCA/Mapbox%20GL%20Cordova%20MBTiles.app/www/index.html
```

And when I connect safari to debug i see the error:  Webcore: Worker(15): EXC_BAD_ACCESS (code=1, address=0x0).

Any idea why this is happening??

**Comment by rklerck** [comment]
Further investigation. It prints out 'before on load' but never the 'before addsourcetype'. it will print out 'ERROR: {"status":0}'.

```
console.log('before on load');
map.on('load', function () {
    console.log('before addsourcetype');
```
    map.addSourceType('mbtiles', mapboxgl.MBTilesSource, function () {

**Comment by oscarfonts** [comment]
So far I haven't had the opportunity to test this project on iOS, as I don't have the hardware to do so.
So I'm afraid I will be of little help here.

Specifically, there is an "if" condition that never was tested, which is what determines where is the mbtiles file being copied:
https://github.com/oscarfonts/mapbox-gl-cordova-mbtiles/blob/master/src/mbtiles_source.js#L42

But it seems that the problem arises even before hitting that line (before adding the source type).

**Comment by rklerck** [comment]
Ive fixed almost everything (will post when it works 100%), but I have still 1 problem, hope you have any ideas. After firing the sql I get this error:

`TypeError: self.atob is not a function. (In 'self.atob(r)', 'self.atob' is undefined)`

Also tried with window instead of self but the result is the same..

**Comment by oscarfonts** [comment]
Wow, very good news!

About self.atob: I've found it's Safari lacking the atob function in web workers:
https://bugs.webkit.org/show_bug.cgi?id=55663

Try using a polyfill such as this one:
https://github.com/davidchambers/Base64.js/blob/master/base64.js

**Comment by rklerck** [comment]
Indeed, that was it! Already figured it out myself.

So, the things I changed:

Use base64-js instead of atob:

```
convertBase64ToUint8Array(base64) {
    return base64js.toByteArray(base64);
}
```

Instead of map.on I used:

`map.once('load', function () {`

Something went wrong with the copy so I used the cordova-plugin-dbcopy plugin to do the copy of the database.

openDatabase works differently on IOS so I did:

`openDatabase({name: name, createFromLocation: 1,location: 2});`

And I didn't use external json files for the style, something went wrong with those. So for now I used variables directly.

And last but not least, the untested IF wasn't working indeed, it should be:

`resolveLocalFileSystemURL(cordova.file.applicationDirectory, resolve, reject);`
